### PR TITLE
[ADD] report_qweb_weasyprint_renderer

### PR DESCRIPTION
**Pull request by hbrunn**
This refers to discussion https://github.com/OCA/reporting-engine/issues/254

For testing this, go to the list view of addons and print the weasyprint demo report.

To make it possible to replace wkhtmltopdf, we'll probably need to implement [box-decoration-break](https://developer.mozilla.org/en-US/docs/Web/CSS/box-decoration-break) in weasyprint

### CODE REVIEW

File: report_qweb_weasyprint_renderer/models/ir_actions_report.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    from weasyprint import HTML
+except ImportError:
+    HTML = None
+
+
+class IrActionsReport(models.Model):
+    _inherit = 'ir.actions.report'
+
+    qweb_pdf_engine = fields.Selection(
+        selection_add=[('weasyprint', 'weasyprint')],
```
Comment (elicoidal) [inline report_qweb_weasyprint_renderer/models/ir_actions_report.py]: nitpicking: Weasyprint?

File: report_qweb_weasyprint_renderer/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2018 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Weasyprint QWEB renderer",
+    "version": "12.0.1.0.0",
+    "author": "Therp BV,Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "category": "Reporting",
+    "summary": "Use weasyprint to create PDFs",
```
Comment (elicoidal) [inline report_qweb_weasyprint_renderer/__manifest__.py]: Weasyprint

**Comment by JordiBForgeFlow** [comment]
Interesting! cc @etobella @ageficent

File: report_qweb_weasyprint_renderer/demo/report.xml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                        <h1 t-attf-id="chapter{{this.id}}" t-esc="this.shortdesc" />
+                        <img class="img" t-attf-src="data:img/*;base64,{{this.icon_image}}" />
+                        <div class="desc" t-esc="this.description" />
+                    </section>
+                </t>
+            </body>
+        </html>
+    </template>
+    <report id="demo_report"
+        string="weasyprint demo report"
```
Comment (elicoidal) [inline report_qweb_weasyprint_renderer/demo/report.xml]: Weasyprint

**Review by elicoidal — APPROVED** [review]
LGTM

**Review by bealdav — APPROVED** [review]
Just tested, it works even if runbot is red

Thanks a lot

**Review by NL66278 — APPROVED** [review]
:+1:  Looks good to me. Nice separation between the general concept of having rendering engines and the weasyprint stuff. Nitpick: we should no longer have copyright lines in __init__.py files.

**Review by etobella — CHANGES_REQUESTED** [review]
It works, but it is not currently showing correctly the headers and footers. I don't know if it is something that I made wrongly or it is the behaviour. I used the runbot and changed the Preview External Report and It was not showing correctly the footer / headers.

File: report_qweb_weasyprint_renderer/models/ir_actions_report.py (hunk 1 above)
Comment (treviser) [reply report_qweb_weasyprint_renderer/models/ir_actions_report.py]: Nitty nitpicking: WeasyPrint? ... https://github.com/Kozea/WeasyPrint

**Comment by hbrunn** [comment]
@etobella this is not a drop in replacement for existing wkhtmltopdf reports (yet), so for testing this, you should use the demo report included here and extend this.

I just did some new experiments, headers and footers work now both with CSS and as fixed position elements, but in the latter case we don't have counters, in the former case no elements, so it's not yet perfect.

**Comment by etobella** [comment]
Then it worked as expected :smile:

**Comment by bealdav** [comment]
While webkit is broken according to versions
see https://github.com/camptocamp/docker-odoo-project/pull/115
weasyprint engine works.

Slowly but works

Then at least, as fallback report engine, this feature should considered as required in our customer projects and PR merge encouraged.

Thanks

**Comment by hbrunn** [comment]
what's missing here is time to finish https://github.com/Kozea/WeasyPrint/pull/882 (=support counters and strings in margin boxes in weasyprint), when this is done, it should be doable to add some CSS magic to use this as dropin replacement for wkhtmltopdf

**Comment by bealdav** [comment]
ok @hbrunn thanks for your info.

**Comment by hbrunn** [comment]
of course, anybody else is cordially invited to finish it too :-)

**Comment by patel1008b** [comment]
@hbrunn @elicoidal

**Comment by maisim** [comment]
@hbrunn @elicoidal @etobella @bealdav 
Hello, it would be nice to merge this work, I thought I would do the same and I find it already done!

**Comment by hbrunn** [comment]
feel free to take over whatever you can use here, I won't be continuing this any time soon

**Comment by maisim** [comment]
Ok, thanks for the work you have done!

**Comment by litmount** [comment]
> @hbrunn @elicoidal @etobella @bealdav Hello, it would be nice to merge this work, I thought I would do the same and I find it already done!

Hello @maisim , Did you make any progress on this? It would be nice to have weasyprint as an option in Odoo.

**Comment by maisim** [comment]
Hi @litmount 
not yet, hopefully in the next few weeks

**Comment by litmount** [comment]
> Hi @litmount not yet, hopefully in the next few weeks

@maisim Great! What version of Odoo are you working on this for?

**Pull request closed without merging**